File: 1190-reverse-substrings-between-each-pair-of-parentheses/1190-reverse-substrings-between-each-pair-of-parentheses.py
class Solution:
    def reverseParentheses(self, s: str) -> str:
        def reverse(sub: list, start: int, end: int):
            while start < end:
                sub[start], sub[end] = sub[end], sub[start]
                start += 1
                end -= 1

        stack = []
        result = []
        for c in s:
            if c == '(':
                stack.append(len(result))
            elif c == ')':
                start = stack.pop()
                reverse(result, start, len(result) - 1)
            else:
                result.append(c)
        
        return ''.join(result)

File: 1190-reverse-substrings-between-each-pair-of-parentheses/1190-reverse-substrings-between-each-pair-of-parentheses.cpp
class Solution {
public:
    string reverseParentheses(string s) {
        stack<int> open;
        string result = "";
        for (char c : s) {
            if (c == '(') {
                int n = result.length();
                open.push(n);
            } else if (c == ')') {
                int start = open.top();
                open.pop();
                reverse(result.begin() + start, result.end());
            } else {
                result += c;
            }
        }
        return result;
    }
};